src/lexer/lexer.h

```diff
@@ -5,30 +5,43 @@
 #include <string>
 
 #include "lexer/token.h"
+#include "error/error.h"
 
 namespace lexer {
+class LexError : public Error {
+ public:
+  explicit LexError(const std::string& message, Location l)
+    : Error(message + "\nAt " + l.toString()) {}
+};
+
 class Lexer {
  public:
   // What type of input is given to the constructor.
   enum class SourceTag {
     FILE,
     STRING,
   };
-  Lexer(const std::string& source, SourceTag tag);
+  Lexer(const std::string& source, SourceTag tag,
+        const std::string& filename = "");
 
   static Lexer from_file(const std::string& file) {
-    return Lexer(file, SourceTag::FILE);
+    return Lexer(file, SourceTag::FILE, file);
   }
 
   static Lexer from_string(const std::string& text) {
-    return Lexer(text, SourceTag::STRING);
+    return Lexer(text, SourceTag::STRING, "<string>");
   }
 
-  Token get_next_token();
+  ErrorOr<Token, LexError> get_next_token();
 
  private:
+  LexError makeError(const std::string& message) const;
+  // Puts the value of the next char into next_char_, and updates the location.
+  MaybeError<LexError> get_next_char();
+
+  char next_char_ = 0;
```

Comment:
Do you plan for the language to be purely ASCII? If you think it will be UTF-8 (like some reasonable languages), this will not work.